Update Truffle to >19.0.1 release

smarr opened the pull request:
This PR gets us onto the latest Truffle version.
There are various little changes, some of which may have bigger impact down the line.

- The interop support is now even more minimal than before, but, it seems sufficient for debugging, and for everything that's currently tested.
- This PR also restore testing on macOS.
- We use now `mx native-image` which simplifies image generation and avoids having to maintain dependencies manually.

@daumayr would be great if you could give this a spin and see whether it works on your machine.

smarr committed:
Update Truffle to pre-vm-20

- also updated mx accordingly

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adapt to frame descriptor and frame slot changes

- account for removal of frame descriptor from slot
- don't use deprecated FrameSlot.getIndex()

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Include jdk1.8 into path

This may need extra support for other JDK versions.

- fixed in build files
- fixed in launcher script

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[Travis] Update JMVCI_VERSION and JDK, and use Trusty

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Address minor deprecations

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Reorder specialization which was indicated as error by Truffle DSL processor

Did not verify

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Work around issue in Truffle DSL

- nested classes do not generate code any more, might be fixed later?
- classes need specializations to generated code, specialization in abstract super class is not sufficient

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] Adapt to new interop system

- implement isNull in new interop
- remove outdated isInstance

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Rework shutdown and entering TruffleContext

- need to enter TruffleContext explicitly on worker threads
  - needed for instrumentation etc

- initiate shutdown explicitly via an eval.
  This avoids races and makes sure the f/j pools are in a known state.
  The drawback of this approach is that we need an extra .eval() call before we can close the context.

- use TruffleContext instead of Context inside of SOMns

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix substrate-related includes

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make JDK download and untar more robust to naming

- top level folder in tar can be named not following expected pattern

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Avoid toString in metrics writer and use specific criteria

This is to avoid breaking tests by Truffle changing some toString format.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update superinstruction results to account for changed And/Or nodes

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added a breakpoint test to actor.ts, easier to debug

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
We have a new suspension point in the STM tests, should fix this eventually

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[Travis] Fix macOS testing

- use xcode9.3 image for Java 8
- install ant

smarr committed:
Add missing initialization to SomPolyglotTests

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix SomPolyglotTests making RootName check more robust

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Avoid issues with ._* files in si and dym expected results

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use mx native-image command to generate som-native

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
InitializerFieldWrite replace neverPartOfCompilation with transferToInterp

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Create Debugging session with root elements enabled

The changes to the debugging session require to use the new support for root elements, otherwise, our root-element-based breakpoints would not work.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@fniephaus possibly relevant for you, too.

@richard-roberts @kjx @IsaacOscar @Jamigrith: this is a major update of Truffle, and we might want to consider this for https://github.com/gracelang/moth-SOMns It allows to use the officially released GraalVM 19

fniephaus commented on src/som/compiler/MixinDefinition.java:
```diff
@@ -73,6 +72,7 @@
  * at runtime, which then also has the super class and mixins resolved to be
  * used to instantiate {@link SClass} objects.
  */
+@ExportLibrary(InteropLibrary.class)
```
Why add `@ExportLibrary(InteropLibrary.class)` to this class if you don't implement any of the interop messages (with `@ExportMessage`)?

smarr replied:
Work in progress

fniephaus replied:
Ok, I was just wondering because this is not a draft pull request. 😉

smarr replied:
This is not going to be addressed in this PR. There's no need for more interop support.

fniephaus commented on src/som/primitives/threading/TaskThreads.java:
```diff
@@ -25,6 +25,7 @@
 
 public final class TaskThreads {
 
+  @ExportLibrary(InteropLibrary.class)
```
Same question (from above).

fniephaus commented on src/som/vmobjects/SObjectWithClass.java:
```diff
@@ -3,12 +3,13 @@
 import som.interpreter.objectstorage.ClassFactory;
 
 
+@ExportLibrary(InteropLibrary.class)
```
And again...

smarr committed:
Move JVMCI download to ant

- remove it from .travis.yml
- use the new libs/jvmci folder in ./som launcher
- remove “known” locations to guess where a GRAAL/JVMCI_BIN is
- adapt order to find JVMCI bin to use more relevant/explicit settings first
- use it also for benchmarks

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix build.xml

Signed-off-by: Stefan Marr <git@stefan-marr.de>
